spec/controllers/newflow/login_signup_controller_spec.rb

```diff
@@ -484,23 +484,47 @@ module Newflow
       end
 
       context 'failure' do
-        before do
-          allow_any_instance_of(OauthCallback).to receive(:mismatched_authentication?).and_return(true)
-        end
+        context 'with mismatched_authentication' do
+          before do
+            allow_any_instance_of(OauthCallback).to receive(:mismatched_authentication?).and_return(true)
+          end
 
-        let(:params) do
-          { provider: 'facebook', uid: 'nonexistent' }
-        end
+          let(:params) do
+            { provider: 'facebook', uid: 'nonexistent' }
+          end
 
-        it 'redirects to login' do
-          get(:oauth_callback, params: params)
-          expect(response).to redirect_to(newflow_login_path)
+          it 'redirects to login' do
+            get(:oauth_callback, params: params)
+            expect(response).to redirect_to(newflow_login_path)
+          end
+
+          it 'saves login failed email' do
+            expect_any_instance_of(described_class).to receive(:save_login_failed_email).and_call_original
+            get(:oauth_callback, params: params)
+            expect(session[:login_failed_email]).to eq(info[:email])
+          end
         end
 
-        it 'saves login failed email' do
-          expect_any_instance_of(described_class).to receive(:save_login_failed_email).and_call_original
-          get(:oauth_callback, params: params)
-          expect(session[:login_failed_email]).to eq(info[:email])
+        context 'when no email address is returned' do
+          let(:info) do
+            { email: nil, name: Faker::Name.name }
+          end
+
+          let(:params) do
+            { provider: 'facebook', uid: Faker::Internet.uuid }
+          end
+
+          before do
+            allow_any_instance_of(OauthCallback).to receive(:oauth_response)  do
+              # TODO: refactor this?
```

Comment:
prob. best to refactor now instead of punting if at all possible.

Reply:
I had a question mark in it because I didn't know if or how I could refactor it. After looking into it, I really don't know how to refactor it 😬 . Removed the comment 🙃 .